### PR TITLE
feat(integration): onboard Facebook Offline Conversions destination

### PR DESCRIPTION
**Pull request by mihir-4116**
## Description of the change
feat(new integration):facebook_offline_conversions cloud mode destination onboarding 
> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [✅ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ✅]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ✅] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/facebook_offline_conversions/utils.js [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+  return [data];
+};
+
+/**
+ * Returns the payload
+ * @param {*} message
+ * @param {*} destination
+ * @returns
+ */
+const offlineConversionResponseBuilder = (message, destination) => {
```
Comment (shrouti1507) [inline v0/destinations/facebook_offline_conversions/utils.js:251]: need to add `limitedDataUsage` facility

File: v0/destinations/facebook_offline_conversions/utils.js [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+/**
+ * Returns eventSetIds and standard event name
+ * @param {*} destination
+ * @param {*} event
+ * @returns
+ */
+const getEventSetId = (destination, event) => {
+  const { Config } = destination;
+  const { eventsToIds, eventsToStandard } = Config;
+  const eventsMapping = getHashFromArray(eventsToStandard, "from", "to", false);
```
Comment (shrouti1507) [inline v0/destinations/facebook_offline_conversions/utils.js:156]: Need to support multiplexing here as well

**Review by shrouti1507 — COMMENTED** [review]
Are we not having any internal event mapping? we can map the following events if we follow Facebook pixel. can we send these events via postman and see if they reflect in the dashboard or not?

```
Products Searched - Search
Product Viewed - ViewContent
Product List Viewed - ViewContent
Product Added to Wishlist - AddToWishlist
Product Added - AddToCart
Checkout Started - InitiateCheckout
Payment Info Entered - AddPaymentInfo
Order Completed - Purchase
```

File: v0/destinations/facebook_offline_conversions/utils.js [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+/**
+ * Returns the data array
+ * @param {*} payload
+ * @param {*} message
+ * @returns
+ */
+const prepareData = (payload, message, destination) => {
+  const { Config } = destination;
+  const { limitedDataUSage, valueFieldIdentifier } = Config;
+  const data = {};
```
Comment (shrouti1507) [inline v0/destinations/facebook_offline_conversions/utils.js:234]:
```
const data = {
    match_keys : prepareMatchKeys(payload),
    event_time :  payload.event_time,
    currency : payload.currency || "USD",
    value :  get(message.properties, valueFieldIdentifier) || payload.value || 0,
    content_type : getContentType(message),
    order_id : payload.order_id || null,
    contents : payload.contents ? getProducts(payload.contents) : null,
    custom_data: message.properties
    ? extractCustomFields(
        message,
        payload,
        ["properties"],
        TRACK_EXCLUSION_FIELDS
      )
    : null
};
```

File: v0/destinations/facebook_offline_conversions/utils.js [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+    delete payload.name;
+  }
+
+  const data = prepareData(payload, message, destination);
+  const eventToIdsArray = getStandardEventsAndEventSetIds(
+    destination,
+    message.event
+  );
+
+  const offlineConversionsPayload = [];
```
Comment (shrouti1507) [inline v0/destinations/facebook_offline_conversions/utils.js:367]: name it as `offlineConversionsPayloads`

File: v0/destinations/facebook_offline_conversions/transform.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+      .setStatTags({
+        destType: DESTINATION,
+        stage: TRANSFORMER_METRIC.TRANSFORMER_STAGE.TRANSFORM,
+        scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.TRANSFORMATION.SCOPE,
+        meta: TRANSFORMER_METRIC.MEASUREMENT_TYPE.TRANSFORMATION.META.BAD_EVENT
+      })
+      .build();
+  }
+
+  const offlineConversionsPayload = offlineConversionResponseBuilder(
```
Comment (shrouti1507) [inline v0/destinations/facebook_offline_conversions/transform.js:50]: `offlineConversionsPayloads`

File: v0/destinations/facebook_offline_conversions/transform.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    message,
+    destination
+  );
+
+  const urls = [];
+  offlineConversionsPayload.forEach(item => {
```
Comment (shrouti1507) [inline v0/destinations/facebook_offline_conversions/transform.js:56]: `offlineConversionsPayloads`

File: v0/destinations/facebook_offline_conversions/utils.js [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+ * @param {*} data
+ * @param {*} ids
+ * @param {*} payload
+ * @returns
+ */
+const prepareUrls = (destination, data, ids, payload) => {
+  const urls = [];
+  const uploadTags = payload.upload_tag || "rudderstack";
+  const [first] = data;
+  const encodedData = `%5B${encodeURIComponent(JSON.stringify(first))}%5D`;
```
Comment (shrouti1507) [inline v0/destinations/facebook_offline_conversions/utils.js:75]: Please add comment, what does this %5B and %5D signify

File: v0/destinations/facebook_offline_conversions/utils.js [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+  });
+
+  return eventSetIds;
+};
+
+/**
+ * Returns eventSetIds and standard event name
+ * @param {*} destination
+ * @param {*} event
+ * @returns
```
Comment (shrouti1507) [inline v0/destinations/facebook_offline_conversions/utils.js:183]: add the response structure here

File: v0/destinations/facebook_offline_conversions/utils.js [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+    }
+  });
+  return products;
+};
+
+/**
+ * Returns actionSource value
+ * @param {*} payload
+ * @returns
+ */
```
Comment (shrouti1507) [inline v0/destinations/facebook_offline_conversions/utils.js:265]: Lets add this ref : https://developers.facebook.com/docs/marketing-api/conversions-api/parameters/server-event#action-source here

File: v0/destinations/facebook_offline_conversions/utils.js [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+          message,
+          payload,
+          ["properties"],
+          TRACK_EXCLUSION_FIELDS
+        )
+      : null,
+    action_source: getActionSource(payload),
+    event_source_url: payload.event_source_url || null
+  };
+  if (limitedDataUSage) {
```
Comment (shrouti1507) [inline v0/destinations/facebook_offline_conversions/utils.js:307]: Lets add this ref: https://developers.facebook.com/docs/marketing-apis/data-processing-options here in comments

File: v0/destinations/facebook_offline_conversions/config.js [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+const ENDPOINT = "https://graph.facebook.com/v15.0/OFFLINE_EVENT_SET_ID/events";
+
+const CONFIG_CATEGORIES = {
+  OFFLINE_EVENTS: {
+    type: "track",
+    name: "FbOfflineConversionsTrackConfig"
+  }
+};
+
+const destKeys = [
```
Comment (shrouti1507) [inline v0/destinations/facebook_offline_conversions/config.js:12]: rename it as `MATCH_KEY_FIELDS`

File: v0/destinations/facebook_offline_conversions/config.js [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+const ENDPOINT = "https://graph.facebook.com/v15.0/OFFLINE_EVENT_SET_ID/events";
+
+const CONFIG_CATEGORIES = {
+  OFFLINE_EVENTS: {
+    type: "track",
+    name: "FbOfflineConversionsTrackConfig"
+  }
+};
+
```
Comment (shrouti1507) [inline v0/destinations/facebook_offline_conversions/config.js:11]: Add this ref: https://developers.facebook.com/docs/marketing-api/offline-conversions/#upload-events here

File: v0/destinations/facebook_offline_conversions/config.js [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  "country",
+  "madid",
+  "extern_id",
+  "lead_id",
+  "fbc",
+  "fbp",
+  "client_user_agent"
+];
+
+const destKeyType = {
```
Comment (shrouti1507) [inline v0/destinations/facebook_offline_conversions/config.js:34]: `MATCH_KEY_FIELD_TYPE_DICTIONARY`

File: v0/destinations/facebook_offline_conversions/transform.js [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+      })
+      .build();
+  }
+
+  const offlineConversionsPayload = offlineConversionResponseBuilder(
+    message,
+    destination
+  );
+
+  const urls = [];
```
Comment (shrouti1507) [inline v0/destinations/facebook_offline_conversions/transform.js:55]: `finalResponseUrls`

File: v0/destinations/facebook_offline_conversions/transform.js [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+const { TRANSFORMER_METRIC } = require("../../util/constant");
+const { offlineConversionResponseBuilder, prepareUrls } = require("./utils");
+const { DESTINATION } = require("./config");
+const { EventType } = require("../../../constants");
+
+const responseBuilder = (endpoint, method) => {
+  if (endpoint) {
+    const response = defaultRequestConfig();
+    response.endpoint = endpoint;
+    response.method = method;
```
Comment (ItsSudip) [inline v0/destinations/facebook_offline_conversions/transform.js:16]: Method will be same for every event. Do we need to send it as a parameter?

**Review by ItsSudip — COMMENTED** [review]
We are doing hashing by default. What if customer is sending hashed data?

**Comment by mihir-4116** [comment]
> We are doing hashing by default. What if customer is sending hashed data?

File: v0/destinations/facebook_offline_conversions/data/FbOfflineConversionsTrackConfig.json [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+  },
+  {
+    "destKey": "event_time",
+    "sourceKeys": "timestamp",
+    "sourceFromGenericMap": true,
+    "required": true,
+    "metadata": {
+      "type": "secondTimestamp"
+    }
+  },
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply v0/destinations/facebook_offline_conversions/data/FbOfflineConversionsTrackConfig.json:21]: @ItsSudip , we are able to send any older timestamp events to offline_conversions. even from segment also able to send it. however, we will mention in doc that event shouldn't be older then 62 days

**Comment by mihir-4116** [comment]
> We are doing hashing by default. What if customer is sending hashed data?

-> confirmed with shrouti, we are not giving option to send hashed data to customer

**Comment by mihir-4116** [comment]
reopening it

**Pull request closed without merging**